dive_core/dive_command_hierarchy.h

```diff
@@ -0,0 +1,130 @@
+    // what the top level root node is
+    DiveVector<uint64_t> m_start_shared_child;
+    DiveVector<uint64_t> m_end_shared_child;
+    DiveVector<uint64_t> m_root_node_index;
+
+    void SetNumNodes(uint64_t num_nodes) override;
+    void AddSharedChildren(uint64_t node_index, const DiveVector<uint64_t> &children);
+};
+
+class DiveCommandHierarchyCreator
+{
+public:
+    DiveCommandHierarchyCreator(CommandHierarchy                  &command_hierarchy,
+                                GfxrVulkanCommandHierarchyCreator &gfxr_command_hierarchy_creator,
+                                CommandHierarchyCreator           &pm4_command_hierarchy_creator);
+
+    bool ProcessDiveSubmits(
+    const DiveVector<SubmitInfo>                                            &submits,
+    const IMemoryManager                                                    &mem_manager,
+    const std::vector<std::unique_ptr<DiveAnnotationProcessor::SubmitInfo>> &gfxr_submits);
+
+    // If flatten_chain_nodes set to true, then chain nodes are children of the top-most
+    // root ib or call ib node, and never a child of another chain node. This prevents a
+    // deep tree of chain nodes when a capture chains together tons of IBs.
+    // Optional: Passing a reserve_size will allow the creator to pre-reserve the memory needed and
+    // potentially speed up the creation
+    bool CreateTrees(DiveCaptureData        &dive_capture_data,
+                     bool                    flatten_chain_nodes,
+                     std::optional<uint64_t> reserve_size);
+
+    void OnCommand(uint32_t submit_index, DiveAnnotationProcessor::VulkanCommandInfo vk_cmd_info);
+
+    void OnGfxrSubmit(uint32_t                                   submit_index,
+                      const DiveAnnotationProcessor::SubmitInfo &submit_info);
+
+    void CreateTopologies();
+
+private:
+    friend class CommandHierarchyCreator;
+    friend class GfxrVulkanCommandHierarchyCreator;
+
+    CommandHierarchy                  &m_command_hierarchy;
+    const Pm4CaptureData              *m_capture_data_ptr = nullptr;
+    const GfxrCaptureData             *m_gfxr_capture_data_ptr = nullptr;
+    GfxrVulkanCommandHierarchyCreator &m_gfxr_command_hierarchy_creator;
+    CommandHierarchyCreator           &m_pm4_command_hierarchy_creator;
+
+    DiveVector<uint64_t> m_node_start_shared_child[CommandHierarchy::kTopologyTypeCount];
+    DiveVector<uint64_t> m_node_end_shared_child[CommandHierarchy::kTopologyTypeCount];
+    DiveVector<uint64_t> m_node_root_node_index[CommandHierarchy::kTopologyTypeCount];
+
+    // This is a list of child indices per node, ie. topology info
+    // Once parsing is complete, we will create a topology from this
+    // There are 2 sets of children per node, per topology. The second set of children nodes can
+    // have more than 1 parent each
+    DiveVector<DiveVector<uint64_t>> m_node_children[CommandHierarchy::kTopologyTypeCount][2];
```

Comment:
sorry i forget what are the [2] here? would it be better to have enum here as well so that it is clear what the 2 elems are?

Reply:
Added a ChildrenNodeType enum that holds the type types: SingleParentNodeChildren and SharedNodeChildren